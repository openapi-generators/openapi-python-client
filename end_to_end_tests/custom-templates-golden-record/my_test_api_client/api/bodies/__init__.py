""" Contains methods for accessing the API Endpoints """

import types

from . import json_like, post_bodies_multiple


class BodiesEndpoints:
    @classmethod
    def post_bodies_multiple(cls) -> types.ModuleType:
        """
        Test multiple bodies
        """
        return post_bodies_multiple

    @classmethod
    def json_like(cls) -> types.ModuleType:
        """
        A content type that works like json but isn't application/json
        """
        return json_like
