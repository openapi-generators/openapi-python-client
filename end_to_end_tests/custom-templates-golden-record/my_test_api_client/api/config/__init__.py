"""Contains methods for accessing the API Endpoints"""

import types

from . import content_type_override


class ConfigEndpoints:
    @classmethod
    def content_type_override(cls) -> types.ModuleType:
        """
        Content Type Override
        """
        return content_type_override
