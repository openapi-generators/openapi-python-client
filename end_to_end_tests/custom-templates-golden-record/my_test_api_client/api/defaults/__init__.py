"""Contains methods for accessing the API Endpoints"""

import types

from . import defaults_tests_defaults_post


class DefaultsEndpoints:
    @classmethod
    def defaults_tests_defaults_post(cls) -> types.ModuleType:
        """
        Defaults
        """
        return defaults_tests_defaults_post
