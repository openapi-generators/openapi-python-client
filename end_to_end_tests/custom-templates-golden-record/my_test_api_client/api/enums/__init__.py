""" Contains methods for accessing the API Endpoints """

import types

from . import bool_enum_tests_bool_enum_post, int_enum_tests_int_enum_post


class EnumsEndpoints:
    @classmethod
    def int_enum_tests_int_enum_post(cls) -> types.ModuleType:
        """
        Int Enum
        """
        return int_enum_tests_int_enum_post

    @classmethod
    def bool_enum_tests_bool_enum_post(cls) -> types.ModuleType:
        """
        Bool Enum
        """
        return bool_enum_tests_bool_enum_post
