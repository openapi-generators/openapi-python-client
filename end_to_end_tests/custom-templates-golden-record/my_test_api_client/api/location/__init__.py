""" Contains methods for accessing the API Endpoints """

import types

from . import get_location_query_optionality


class LocationEndpoints:
    @classmethod
    def get_location_query_optionality(cls) -> types.ModuleType:
        return get_location_query_optionality
