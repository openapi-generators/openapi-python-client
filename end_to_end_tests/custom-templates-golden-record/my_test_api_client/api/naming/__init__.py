""" Contains methods for accessing the API Endpoints """

import types

from . import post_naming_property_conflict_with_import


class NamingEndpoints:
    @classmethod
    def post_naming_property_conflict_with_import(cls) -> types.ModuleType:
        return post_naming_property_conflict_with_import
