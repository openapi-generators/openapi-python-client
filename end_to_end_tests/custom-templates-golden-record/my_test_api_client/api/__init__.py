""" Contains methods for accessing the API """

from typing import Type

from my_test_api_client.api.default import DefaultEndpoints
from my_test_api_client.api.parameters import ParametersEndpoints
from my_test_api_client.api.tests import TestsEndpoints


class MyTestApiClientApi:
    @classmethod
    def tests(cls) -> Type[TestsEndpoints]:
        return TestsEndpoints

    @classmethod
    def default(cls) -> Type[DefaultEndpoints]:
        return DefaultEndpoints

    @classmethod
    def parameters(cls) -> Type[ParametersEndpoints]:
        return ParametersEndpoints
