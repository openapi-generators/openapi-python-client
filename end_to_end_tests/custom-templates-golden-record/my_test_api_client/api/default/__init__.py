""" Contains methods for accessing the API Endpoints """

import types

from my_test_api_client.api.default import get_common_parameters, post_common_parameters


class DefaultEndpoints:
    @classmethod
    def get_common_parameters(cls) -> types.ModuleType:
        return get_common_parameters

    @classmethod
    def post_common_parameters(cls) -> types.ModuleType:
        return post_common_parameters
