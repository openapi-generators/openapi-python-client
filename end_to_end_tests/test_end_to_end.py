import shutil
from filecmp import cmpfiles, dircmp
from pathlib import Path
from typing import Dict, List, Optional

import pytest
from typer.testing import CliRunner

from openapi_python_client.cli import app


def _compare_directories(
    record: Path,
    test_subject: Path,
    expected_differences: Dict[Path, str],
    depth=0,
):
    """
    Compare two directories and assert that only expected_differences are different

    Args:
        record: Path to the expected output
        test_subject: Path to the generated code being checked
        expected_differences: key: path relative to generated directory, value: expected generated content
        depth: Used to track recursion
    """
    first_printable = record.relative_to(Path.cwd())
    second_printable = test_subject.relative_to(Path.cwd())
    dc = dircmp(record, test_subject)
    missing_files = dc.left_only + dc.right_only
    if missing_files:
        pytest.fail(f"{first_printable} or {second_printable} was missing: {missing_files}", pytrace=False)

    expected_differences = expected_differences or {}
    _, mismatches, errors = cmpfiles(record, test_subject, dc.common_files, shallow=False)
    mismatches = set(mismatches)

    expected_path_mismatches = []
    for file_name in mismatches:
        mismatch_file_path = test_subject.joinpath(file_name)
        expected_content = expected_differences.get(mismatch_file_path)
        if expected_content is None:
            continue

        generated_content = (test_subject / file_name).read_text()
        assert generated_content == expected_content, f"Unexpected output in {mismatch_file_path}"
        expected_path_mismatches.append(mismatch_file_path)

    for path_mismatch in expected_path_mismatches:
        matched_file_name = path_mismatch.name
        mismatches.remove(matched_file_name)
        del expected_differences[path_mismatch]

    if mismatches:
        pytest.fail(
            f"{first_printable} and {second_printable} had differing files: {mismatches}, and errors {errors}",
            pytrace=False,
        )

    for sub_path in dc.common_dirs:
        _compare_directories(
            record / sub_path, test_subject / sub_path, expected_differences=expected_differences, depth=depth + 1
        )

    if depth == 0 and len(expected_differences.keys()) > 0:
        failure = "\n".join([f"Expected {path} to be different but it was not" for path in expected_differences.keys()])
        pytest.fail(failure, pytrace=False)


def run_e2e_test(extra_args: List[str], expected_differences: Dict[Path, str]):
    runner = CliRunner()
    openapi_path = Path(__file__).parent / "openapi.json"
    config_path = Path(__file__).parent / "config.yml"
    gr_path = Path(__file__).parent / "golden-record"
    output_path = Path.cwd() / "my-test-api-client"
    shutil.rmtree(output_path, ignore_errors=True)

    args = ["generate", f"--config={config_path}", f"--path={openapi_path}"]
    if extra_args:
        args.extend(extra_args)
    result = runner.invoke(app, args)

    if result.exit_code != 0:
        raise result.exception

    # Use absolute paths for expected differences for easier comparisons
    expected_differences = {output_path.joinpath(key): value for key, value in expected_differences.items()}
    _compare_directories(gr_path, output_path, expected_differences=expected_differences)

    import mypy.api

    out, err, status = mypy.api.run([str(output_path), "--strict"])
    assert status == 0, f"Type checking client failed: {out}"

    shutil.rmtree(output_path)


def test_end_to_end():
    run_e2e_test([], {})


def test_custom_templates():
    expected_differences = {}  # key: path relative to generated directory, value: expected generated content
    expected_difference_paths = [
        Path("README.md"),
        Path("my_test_api_client").joinpath("api", "__init__.py"),
        Path("my_test_api_client").joinpath("api", "tests", "__init__.py"),
        Path("my_test_api_client").joinpath("api", "default", "__init__.py"),
        Path("my_test_api_client").joinpath("api", "parameters", "__init__.py"),
    ]

    golden_tpls_root_dir = Path(__file__).parent.joinpath("custom-templates-golden-record")
    for expected_difference_path in expected_difference_paths:
        expected_differences[expected_difference_path] = (golden_tpls_root_dir / expected_difference_path).read_text()

    run_e2e_test(
        extra_args=["--custom-template-path=end_to_end_tests/test_custom_templates/"],
        expected_differences=expected_differences,
    )
