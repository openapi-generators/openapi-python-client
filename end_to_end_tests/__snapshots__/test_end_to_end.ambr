# serializer version: 1
# name: test_documents_with_errors[circular-body-ref]
  '''
  Generating /test-documents-with-errors
  Warning(s) encountered while generating. Client was generated, but some pieces may be missing
  
  WARNING parsing POST / within default. Endpoint will not be generated.
  
  Circular $ref in request body
  
  
  If you believe this was a mistake or this tool is missing a feature you need, please open an issue at https://github.com/openapi-generators/openapi-python-client/issues/new/choose
  
  '''
# ---
# name: test_documents_with_errors[missing-body-ref]
  '''
  Generating /test-documents-with-errors
  Warning(s) encountered while generating. Client was generated, but some pieces may be missing
  
  WARNING parsing POST / within default. Endpoint will not be generated.
  
  Could not resolve $ref #/components/requestBodies/body in request body
  
  
  If you believe this was a mistake or this tool is missing a feature you need, please open an issue at https://github.com/openapi-generators/openapi-python-client/issues/new/choose
  
  '''
# ---
# name: test_documents_with_errors[optional-path-param]
  '''
  Generating /test-documents-with-errors
  Warning(s) encountered while generating. Client was generated, but some pieces may be missing
  
  WARNING parsing GET /{optional} within default. Endpoint will not be generated.
  
  Path parameter must be required
  
  Parameter(name='optional', param_in=<ParameterLocation.PATH: 'path'>, description=None, required=False, deprecated=False, allowEmptyValue=False, style=None, explode=False, allowReserved=False, param_schema=Schema(title=None, multipleOf=None, maximum=None, exclusiveMaximum=None, minimum=None, exclusiveMinimum=None, maxLength=None, minLength=None, pattern=None, maxItems=None, minItems=None, uniqueItems=None, maxProperties=None, minProperties=None, required=None, enum=None, const=None, type=<DataType.STRING: 'string'>, allOf=[], oneOf=[], anyOf=[], schema_not=None, items=None, properties=None, additionalProperties=None, description=None, schema_format=None, default=None, nullable=False, discriminator=None, readOnly=None, writeOnly=None, xml=None, externalDocs=None, example=None, deprecated=None), example=None, examples=None, content=None)
  
  If you believe this was a mistake or this tool is missing a feature you need, please open an issue at https://github.com/openapi-generators/openapi-python-client/issues/new/choose
  
  '''
# ---
